Release guide (#72)
spacetelescope/style-guides: guides/release-workflow.md

> # Release And Maintain a Package Using GIT

Reviewer: ```suggestion
# Release And Maintain a Package Using Git
```

## Good Practicies

There are certain activities which are common for all workflows and are simply good practicies when releasing a package.

1. Version the package using semantic versioning.
   Details can be found in the
   [versioning guide ](https://github.com/spacetelescope/style-guides/blob/master/guides/software-versioning.md).

   Before a release update the version of the package to the intended version number. After a release is tagged in the repository and before going back to normal development change the version string in setup.py to the next release, followed by dev. For example, if v 2.1.3 is to be released and the next version will be 2.2.0, before a release is made, set the version string in setup.py to
   version = “2.1.3”. After the release is done, set the version string to version=”2.2.0.dev”.

   [PEP440](https://www.python.org/dev/peps/pep-0440) describes python package versioning conventions.

2. Maintain a change log for the package.

>    It is useful to users if the package has a change log. Many packages maintain a file at the top level, called CHANGES.rst. This file should have a section for each released version and possibly subsections for new features, bug fixes and API changes. Examples cna be found in [astropy](https://github.com/astropy/astropy/blob/master/CHANGES.rst) and [jwst](https://github.com/spacetelescope/jwst/blob/master/CHANGES.rst).

Reviewer: ```suggestion
   It is useful to users if the package has a change log. Many packages maintain a file at the top level, called CHANGES.rst. This file should have a section for each released version and possibly subsections for new features, bug fixes and API changes. Examples can be found in [astropy](https://github.com/astropy/astropy/blob/master/CHANGES.rst) and [jwst](https://github.com/spacetelescope/jwst/blob/master/CHANGES.rst).
```

3. When preparing a release use branches and pull requests (PR) following the [Workflow for contributing to a git](https://github.com/spacetelescope/style-guides/blob/master/guides/git-workflow.md).

4. Tag the software with a "release tag" in the repository.

   Create an annotated tag ("release tag") in the repository for each release. The tag should be associated with the branch the release is made from.
   The command to do this is:

   ```
   git checkout <release_branch>

>    git tag -a "2.1.3" -m "Tagging release v 2.1.3"

Reviewer 2: It's pretty standard to have a "v" in the tag name, e.g. "v2.1.3".  Was that omission intentional?

+1 for annotated tags.

>    git tag -a "2.1.3" -m "Tagging release v 2.1.3"

Reviewer 3: @jhunkeler has reasons not to like the "v" prefix.

>    git tag -a "2.1.3" -m "Tagging release v 2.1.3"

Reviewer 4: The canonical semantic versioning specification (https://semver.org/) mentioned the 'v' prefix in version `1.0.0`, (SemVer tagging), but this was removed in semver 2.0.0.

   ```

5. Publishing a Release on Github

   Online, at the repository root, click on “releases” → “Draft a New Release”.
   Enter the release tag. Describe the release, enter highlights and publish it.

6. Release on PyPi

>    It is common that releases of python packages are published on the Python Packaging Index (PyPi) site. This allows users to find them and install them using `pip`. A PyPi account is required.

Reviewer 4: Since the software in `spacetelescope` is being centrally managed to an extent by virtue of it being hosted in that organization, it makes sense to also apply the same controls to how the software is packaged. Astroconda has a central point of control, but PyPI does not. A management account added to each PyPI package would allow publication control, documentation updates, etc, in the same (possibly automated) manner.

This also acknowledges that it is common to publish via PyPI, but are there requirements to do so for particular types of projects? What determines which packages get published there vs. Astroconda vs. conda-forge vs. other places?

>    It is common that releases of python packages are published on the Python Packaging Index (PyPi) site. This allows users to find them and install them using `pip`. A PyPi account is required.

Author: I can imagine how a scientist may have a package solving a specific issue which they want to manage and make available to others. It doesn't have to be in astroconda but having it on PyPi would be useful. So I think in many cases it's up to the author. But I agree that having an STScI account on the PyPi project page in addition to the author's account would be desirable. However, this is probably more appropriate for a policy. i'm not sure it has a place here, though I can mention it . Whose account do we use in this case?

>    It is common that releases of python packages are published on the Python Packaging Index (PyPi) site. This allows users to find them and install them using `pip`. A PyPi account is required.

Reviewer 4: There is a PyPI account `stsci_ssb` which has been added as a maintainer to many, but not all spacetelescope packages. I don't know who has the credentials to it, but I suspect @jhunkeler is one of them.

>    It is common that releases of python packages are published on the Python Packaging Index (PyPi) site. This allows users to find them and install them using `pip`. A PyPi account is required.

Reviewer 5: the same is true for readthedocs, we've added an stsci maintainer account which should have admin rights on all the stsci owned packages.

   For publishing a release of a python package on PyPi it's best to use [twine](https://pypi.org/project/twine/) - a utility for publishing Python packages on PyPi. Not surprisingly `twine` itself is published on PyPi and can be installed using `pip`.
   The commands for publishing a release on PyPi are:

   ```
   $ git checkout <release_tag>
   $ git clean -fxd
   $ python setup.py build sdist --format=gztar
   $ twine upload dist/*
   ```

   `Note:` The `Project Description` on PyPi is populated by the `Classifiers` in setup.py. It is useful to populate those accurately and completely. The `long dscription` field is used to populate the "Project Description" on PyPi.

## Specific release procedures

> The following two sections provide specific instructions for releasing a python package in the two main use cases listed above. it is assumed the person doing the release uses the spacetelescope repository directly as a working directory. This means the main repository at spaetelescope nneds to be cloned and that "origin" points to spacetelescope, not to the person's fork.

Reviewer: ```suggestion
The following two sections provide specific instructions for releasing a python package in the two main use cases listed above. it is assumed the person doing the release uses the `spacetelescope` repository directly as a working directory. This means the main repository at `spacetelescope` needs to be cloned and that "origin" points to `spacetelescope`, not to the person's fork.
```

  ```
  $ git clone git@github.com:spacetelescope/mypackage.git
  $ git remote -v
    origin git@github.com:spacetelescope/mypackage.git (fetch)
    origin git@github.com:spacetelescope/mypackage.git (pull)
  ```

### Developing and releasing from the master branch.

In this case all work is done on the `master` branch. Old releases are not maintained. When the code is ready for release, set the version in setup.py and set the date of the release in CHANGES.rst. Tag the release on master.

  ```
  git checkout master
  git status
  On branch master
  ```

Create an annotated tag which corresponds to the release and push it to the repository. Publish the release on github and PyPi.

### Use a different release branch for each release.

> In this case the release is prepared on a dedicated branch ("release branch"). It is possible to maintain old releases by doing a patch release off a release branch. The instructions below use a third branch called "stable". It is a branch where final releases land and are tagged.

Author: It is the "production" branch. All release branches get merged with it and the final tag is on the stable branch. It is not strictly necessary but is a convenience and that's how we do it for jwst and some other projects. I will clarify this a bit in the guide.

The specific steps are:

1. Start by editing the change log file (and possibly setup.py) to set the release version and date. Do this on the master branch.

2. Prepare a release branch.

   Make sure you are on the master branch. Make a branch for the release. Name the branch 2.1.3x. Always use a letter at the end of the branch name so that the version number (2.1.3) can be used later as a tag as duplicae names are harmful.

   ```
   $ git remote -v
     origin git@github.com:spacetelescope/jwst.git (fetch)
     origin git@github.com:spacetelescope/jwst.git (push)

   $ git status # should show you are on master.
     On branch master
     Your branch is up-to-date with 'origin/master'.
     nothing to commit, working tree clean

   $ git branch 2.1.3x
   $ git checkout 2.1.3x
   ```

> 3. Push the 2.1.3x branch to spacetelescope and create a PR to the "stable" branch. Note, that origin below refers to the spacetelescope repository. When tests pass merge with "stable".

Reviewer: ```suggestion
3. Push the 2.1.3x branch to `spacetelescope` and create a PR to the "stable" branch. Note, that origin below refers to the `spacetelescope` repository. When tests pass merge with "stable".
```

Author: `stable is a branch in astropy as well. Can you clarify what exactly is confusing?

   `git push origin 2.1.3x`

4. Checkout and update the "stable" branch and tag it with the new release.

   ```
   $ git checkout stable
   $ git pull origin stable
   $ git tag -a "2.1.3" -m "tagging release 2.1.3"
   $ git push origin 2.1.3
   ```

   Prepare master for further development

5. Update the change log for the next release and push it to master. Make sure the current branch is "master".

>    Note: If this package is to be installed nightly internally in conda-dev then "master" needs to be tagged with a "newer" tag which is picked up by conda-dev build scripts. The suggested tag is the next version, followed by a. The tag name should be chosen in such a way that conda would resolve it as the latest tag. This allows the nightly build scripts to pick up the latest version from master.

Author: This is not clear, isn't it? A character is appended to the tag name and the suggestion is to use the character `a` which is officially recognized by PEP 440. So if the release is tagged with `1.2.3` and the next release will be `1.3.0` the tag should be `1.3.0a`. I will clarify this in the guide.

>    Note: If this package is to be installed nightly internally in conda-dev then "master" needs to be tagged with a "newer" tag which is picked up by conda-dev build scripts. The suggested tag is the next version, followed by a. The tag name should be chosen in such a way that conda would resolve it as the latest tag. This allows the nightly build scripts to pick up the latest version from master.

Author: no

# Release And Maintain a Package Using Git

## Introduction

This document aims to facilitate the process of releasing a package by providing
good practices and possible workflows and instructions for managing a release for 
a code repository. Although the document focuses on and uses python packages as 
examples, many of the “good practices” are general and apply to software using any

> language. Two kindsof workflow are discussed. Which one to use depends on whether 

Reviewer 3: ```suggestion
language. Two kinds of workflow are discussed. Which one to use depends on whether 
```
